fl4health/losses/mkmmd_loss.py

```diff
@@ -77,156 +80,218 @@ def construct_quadruples(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor
+
+        # each inner product is a tensor of dimension len(v_i_quadruples), we return a tensor of shape
+        # len(v_i_quadruples) x 4
+        return torch.cat([x_x_prime, y_y_prime, x_y_prime, x_prime_y], dim=1)
+
     def compute_h_u_from_inner_products(self, inner_products: torch.Tensor, gamma: torch.Tensor) -> torch.Tensor:
         # Gamma should be of shape torch.Tensor([gamma])
         assert gamma.shape == (1,)
-        # inner_products has shape number of 4 x v_i_quadruples x v_i_quadruples or 4 x len(X) x len(Y)
-        # h_u_componets should have the same shape
+        # inner_products has shape of 4 x n_samples x n_samples
+        # h_u_components should have the same shape
+        h_u_components = torch.exp((-1 * inner_products) / gamma)
+        # Each first dimension of h_u_components should now be u(x_j, x_k), u(y_j, y_k), u(x_j, y_k), and u(y_j, x_k),
+        # where u is the kernel_index^th kernel and j, k index over samples
+        # So we compute:
+        #   h_u[x_j, x_k,y_j, y_k] =  u(x_j, x_k) + u(y_j, y_k) - u(x_j, y_k) - u(y_j, x_k)
+        h_u = h_u_components[0, :] + h_u_components[1, :] - h_u_components[2, :] - h_u_components[3, :]
+        # this results in a matrix of shape 1 x n_samples x n_samples
+        return h_u.unsqueeze(0)
+
+    def compute_h_u_from_inner_products_linear(
+        self, inner_products: torch.Tensor, gamma: torch.Tensor
+    ) -> torch.Tensor:
+        # Gamma should be of shape torch.Tensor([gamma])
+        assert gamma.shape == (1,)
+        # inner_products has shape number of len(v_i_quadruples) x 4 since this is the linear approximation strategy
+        # h_u_components should have the same shape
         h_u_components = torch.exp((-1 * inner_products) / gamma)
         # Each column of h_u_components should now be u(x_{2i-1}, x_{2i}), u(y_{2i-1}, y_{2i}), u(x_{2i-1}, y_{2i}),
-        # and u(x_{2i}, y_{2i-1}), where u is the kernel_index^th kernel
+        # and u(x_{2i}, y_{2i-1}), where u is the kernel_index^th kernel and i indexes over quadruples
         # So we compute:
         #   h_u[x_{2i-1}, x_{2i},y_{2i-1}, y_{2i}] =  u(x_{2i-1}, x_{2i}) + u(y_{2i-1}, y_{2i})
-        # - u(x_{2i-1}, y_{2i}) - u(x_{2i}, y_{2i-1})
-        h_u = h_u_components[0, :] + h_u_components[1, :] - h_u_components[2, :] - h_u_components[3, :]
-        # this results in a matrix of shape 1 xlen(v_i_quadruples) x len(v_i_quadruples) or 1 x len(X) x len(Y)
+        #       - u(x_{2i-1}, y_{2i}) - u(x_{2i}, y_{2i-1})
+        h_u = h_u_components[:, 0] + h_u_components[:, 1] - h_u_components[:, 2] - h_u_components[:, 3]
+        # this results in a matrix of shape 1 x number of v_i_quadruples
         return h_u.unsqueeze(0)
 
-    def compute_all_h_u_from_inner_products(self, inner_product_quadruples: torch.Tensor) -> torch.Tensor:
+    def compute_all_h_u_from_inner_products(self, inner_product_all_samples: torch.Tensor) -> torch.Tensor:
```

Comment:
`compute_all_h_u_from_inner_products` and `compute_all_h_u_from_inner_products_linear` actually have the same flow, but they act on different shaped tensors. So I broke them apart to allow for detailed comments.

Reply:
That is true, thanks for correcting it. I think otherwise I would have get errors for running with linear approximation.

Reply:
Maybe. I think you were handling most stuff properly actually. I just thought it might be easier to understand with some specific comments given that the structures are a little different.